### PR TITLE
Cdp 450 banner notifications

### PR DESCRIPTION
**Pull request by b-pmcg**

### CODE REVIEW

File: src/components/CDPDisplay/Presentation.js [hunk 1]
```diff
@@ -56,6 +58,24 @@ export default function({ cdp, showSidebar, account, network }) {
...
+  //     showBanner({
+  //       banner: 'claim',
+  //       props: {
+  //         colName: 'testName',
+  //         amount: cdp.unlockedCollateral.toNumber(),
+  //         symbol: 'SYM'
+  //       }
+  //     });
+  //   }
+  // }, [cdp, showBanner, unlockedCollateral]);
```
Comment (b-pmcg) [inline src/components/CDPDisplay/Presentation.js]: I thought adding the dependencies here would prevent the re-render issues, when `showBanner` is called it changes the DOM and triggers a render, which calls `showBanner` again, and so on.

File: src/pages/Overview.js [hunk 2]
```diff
@@ -114,10 +114,11 @@ function Overview() {
...
-      showBanner({
-        banner: 'claim',
-        props: { colName: 'NAME', amount: '5.55', symbol: 'SYM' }
-      });
+      // TODO: when uncommented, this causes far to many re-renders, but does not cause infinite loop
+      // showBanner({
+      //   banner: 'claim',
+      //   props: { colName: 'NAME', amount: '5.55', symbol: 'SYM' }
+      // });
     }
```
Comment (b-pmcg) [inline src/pages/Overview.js]: Calling `showBanner` here triggers a lot of renders, but it doesn't break the page like in `Presentation.js`

File: src/components/Banners/CollateralClaim.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+const CollateralClaim = ({ colName, amount, symbol }) => {
+  const { lang } = useLanguage();
+  const { maker, newTxListener } = useMaker();
+
+  const frobToOwnUrn = async () => {
+    //TODO need to fix re-render infinite loop to pass in real props
+    const cdpId = 183;
+    const amount = 0.08;
+
```
Comment (b-pmcg) [inline src/components/Banners/CollateralClaim.js]: note these are hardcoded at the moment, so you can view [CDP 183](http://localhost:3000/borrow/183?network=kovan) to repro the issue (or hardcode your own values)

**Comment by levity** [comment]
To fix your specific issue with infinite re-renders, I would add a property to useBanner's return value indicating what's already being shown, and not call showBanner again if it's already shown.

Zooming out a second, I don't think we need to have all these providers. This hook (and useToggle and probably others) could just be implemented using the StoreProvider.

**Comment by levity** [comment]
Oh, I see now that the infinite re-render problem might be happening because you're causing a side-effect from the `useMemo` callback. This is not what it's for. It's for memoizing expensive calculations that do not have side effects.

`useEffect` runs only after the current render completes; `useMemo` has no such guaranteed behavior. 

So doing something that triggers a new render right in the middle of rendering is probably the root problem here.

For example, perhaps useMemo is implemented so that it saves its memoized value only after the render finishes. (This is certainly possible, but just a guess; I'm not going to go digging in the React code now to figure it out though.)

**Comment by sirromdev** [comment]
@levity @b-pmcg 

I've made a few changes over the weekend as I'm flying out in a few hours and won't be too active on Monday. I've taken on your advice Lawrence and implemented a `useNotification()` hook to use the store rather than use a new provider. I have changed around a few things but if the changes I have made are not to your liking, by all means, revert them back.

https://github.com/makerdao/mcd-cdp-portal/pull/127/files#diff-393e2092f722b579b1bbe9850772bb03R69-R81

The above is an example case (which won't trigger re-renders) where we can add a notification while Presentation.js is mounted and trigger the removal of the notification when it is unmounted. However, if the guardrail if statements are removed we will get infinite re-renders.

Notifications are keyed by an `id` attribute under `notifications.banners` with values being whatever props we want to pass into `components/Notifications.js`. The `useNotification` hook just serves as cleanup instead of writing out wordy dispatch calls wherever we wish to add one.

**Comment by b-pmcg** [comment]
@levity It turns out the excessive re-renders were caused by including properties from the custom hook in the `useEffect` block and adding them as dependencies. Removing them from the dependency array fixed the issue-- which was annoying because eslint exhaustive deps was auto-adding them to the array on save 😱

@sirromdev I appreciate your refactor, but I was running into the same issues as before due to the dependency bug, so I finished CDP-477 using the provider approach and merged it into develop. Feel free to include your changes as you work on CDP-450.

**Comment by levity** [comment]
Ahhhhh because `show` is being redefined every time in the render method of BannerProvider, so it's never the same between re-renders. I think you could use `useCallback` to fix this

**Comment by levity** [comment]
> Ahhhhh because `show` is being redefined every time in the render method of BannerProvider, so it's never the same between re-renders. I think you could use `useCallback` to fix this

@b-pmcg @sirromdev just making sure you saw this

**Pull request merged**